oauth2_http/java/com/google/auth/oauth2/S2A.java

```diff
@@ -0,0 +1,112 @@
+package com.google.auth.oauth2;
+
+import com.google.api.client.http.GenericUrl;
+import com.google.api.client.http.HttpRequest;
+import com.google.api.client.http.HttpResponse;
+import com.google.api.client.json.JsonObjectParser;
+import com.google.api.client.util.GenericData;
+import com.google.auth.http.HttpTransportFactory;
+import com.google.common.collect.Iterables;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.ServiceLoader;
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * Utilities to fetch the S2A (Secure Session Agent) address from the mTLS configuration.
+ *
+ * <p>mTLS configuration is queried from the MDS MTLS Autoconfiguration endpoint.
+ */
+@ThreadSafe
+public final class S2A {
+  public static final String MTLS_CONFIG_ENDPOINT =
+      "/computeMetadata/v1/instance/platform-security/auto-mtls-configuration";
+
+  public static final String METADATA_FLAVOR = "Metadata-Flavor";
+  public static final String GOOGLE = "Google";
+  private static final int MAX_MDS_PING_TRIES = 3;
+  private static final String PARSE_ERROR_S2A = "Error parsing Mtls Auto Config response.";
+
+  private MtlsConfig config;
+
+  private transient HttpTransportFactory transportFactory;
+
+  public S2A() {}
+
+  public void setHttpTransportFactory(HttpTransportFactory tf) {
+    this.transportFactory = tf;
+  }
+
+  /** @return the mTLS S2A Address from the mTLS config. */
+  public synchronized String getMtlsS2AAddress() {
+    if (config == null) {
+      config = getMdsMtlsConfig();
+    }
+    return config.getMtlsS2AAddress();
+  }
+
+  /** @return the plaintext S2A Address from the mTLS config. */
+  public synchronized String getPlaintextS2AAddress() {
+    if (config == null) {
+      config = getMdsMtlsConfig();
+    }
+    return config.getPlaintextS2AAddress();
+  }
+
+  /**
+   * Queries the MDS mTLS Autoconfiguration endpoint and returns the {@link MtlsConfig}.
+   *
+   * <p>Returns {@link MtlsConfig} with empty addresses on error.
+   *
+   * @return the {@link MtlsConfig}.
+   */
+  private MtlsConfig getMdsMtlsConfig() {
+    String plaintextS2AAddress = "";
+    String mtlsS2AAddress = "";
+
+    String url = getMdsMtlsEndpoint();
+    GenericUrl genericUrl = new GenericUrl(url);
+    
+    for (int i = 0; i < MAX_MDS_PING_TRIES; i++) {
+      try {
+        if (transportFactory == null) {
+          transportFactory =
+              Iterables.getFirst(
+                  ServiceLoader.load(HttpTransportFactory.class), OAuth2Utils.HTTP_TRANSPORT_FACTORY);
+        }
+        HttpRequest request =
+            transportFactory.create().createRequestFactory().buildGetRequest(genericUrl);
+        JsonObjectParser parser = new JsonObjectParser(OAuth2Utils.JSON_FACTORY);
```

Comment:
Should this be outside of the loop?

Reply:
Done in https://github.com/googleapis/google-auth-library-java/pull/1400/commits/05aa9cc3d33b57e2d06541f51011c6b850286719